util/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filewatcher
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"k8s.io/klog/v2"
+)
+
+var watchCertificateFileOnce sync.Once
+
+// WatchFileForChanges watches the file, fileToWatch, for changes. If the file contents have changed, the pod this
+// function is running on will be restarted.
+func WatchFileForChanges(fileToWatch string) error {
+	var err error
+
+	// This starts only one occurrence of the file watcher, which watches the file, fileToWatch.
+	watchCertificateFileOnce.Do(func() {
+		klog.V(2).Infof("Starting the file change watcher on file, %s", fileToWatch)
+
+		// Update the file path to watch in case this is a symlink
+		fileToWatch, err = filepath.EvalSymlinks(fileToWatch)
+		if err != nil {
+			return
+		}
+		klog.V(2).Infof("Watching file, %s", fileToWatch)
+
+		// Start the file watcher to monitor file changes
+		go func() {
+			err := checkForFileChanges(fileToWatch)
+			klog.Errorf("Error checking for file changes: %v", err)
+		}()
+	})
+	return err
+}
+
+// checkForFileChanges starts a new file watcher. If the file is changed, the pod running this function will exit.
+func checkForFileChanges(path string) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if ok && (event.Has(fsnotify.Write) || event.Has(fsnotify.Chmod) || event.Has(fsnotify.Remove)) {
+					klog.Infof("file, %s, was modified, exiting...", event.Name)
+					os.Exit(0)
```

Comment:
Is there some more graceful way we can handle this? I don't remember off the top of my head what kinds of cleanup things we might be doing now in main (or will need to do in the future), but I'd prefer not to have to worry about that at all 😄.

Reply:
I'm open to alternative suggestions. This is just the pattern we were using in a different project for this specific scenario with Azure SDK not being able to hotload the new cert and requiring the pod to restart in order to use the new cert.

Maybe I'm missing it but I don't see any cleanup things in the main.go. Is that where you were referencing or was it someplace else.

Reply:
> Azure SDK doesn't support hotloading a new certificate yet

What exactly does this mean? CAPZ orchestrates reading the cert from disk and using that to create a `TokenCredential`, and it's already doing that once in every reconciliation. Do successive calls to `azidentity.NewClientCertificateCredential` with different certificates not produce working creds?

Reply:
@nojnhuh sorry for the delay, I was doing some testing related this I wanted to check first.

So, if we are grabbing the certificate and re-authenticating with Azure, we shouldn't need this filewatcher to reboot the pod. Like:
```
parsedCertificate, key, err := azidentity.ParseCertificates(certsContent, nil)
	if err != nil {
		return err
	}

	options := &azidentity.ClientCertificateCredentialOptions{
		SendCertificateChain: true,
	}
	creds, err := azidentity.NewClientCertificateCredential(tenantID, options, ... etc.)
```

However, with #5283, when something like Secrets Store CSI updates the certificate in the CAPZ pod, that should invalidate the cache right? The invalidated cache would then trigger a re-authentication with Azure with a new certificate?

Reply:
Yes, that's what should happen.

Reply:
If you could validate though that would be great!

Reply:
Ok I'll pull your PR tomorrow and make an image out of it and give it a go.

Reply:
tl;dr - I'm going to close this PR. I do not think we need it given the caching PRs mentioned previously in this thread.

I created an image off the latest in CAPZ to use in a deployment with a sidecar container to check to see CAPZ didn't error out when the certificates rotated. In my project, I just modified the deployment of CAPZ and added a sidecar container and a shared volume between the two containers. I was able to exec into that pod, see the certificate, manually rotate the certificate in key vault and disable the old certificate, and see the new certificate in the sidecar container.

The CAPZ pod never rebooted, there were no errors in the log, and I was able to successfully scale up my nodepool using the new certificate.